File: backend/app/routes/coze.py
"""
Coze数据上传API路由
"""

from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks
from typing import List, Optional
from uuid import UUID
import logging
from sqlalchemy.ext.asyncio import AsyncSession

from app.schemas.coze import (
    CozeUploadRequest,
    CozeUploadResponse,
    CozeTableInfo,
    CozeUploadHistory,
    CozeUploadStatus,
    CozeSyncConfigResponse,
    CozeSyncConfigListResponse
)
from app.services.coze_service import CozeService
from app.services.operation_log_service import OperationLogService
from app.core.database import get_async_db
from app.utils.mapper import snake_to_camel, camel_to_snake  # 添加命名转换工具导入

router = APIRouter()
logger = logging.getLogger(__name__)


@router.get("/coze/tables", response_model=List[CozeTableInfo])
async def get_available_tables(db: AsyncSession = Depends(get_async_db)):
    """获取可上传的数据表列表"""
    try:
        tables = await CozeService.get_available_tables(db)
        # 应用命名转换，确保返回数据符合前端期望的命名格式
        return snake_to_camel(tables)
    except Exception as e:
        logger.error(f"获取数据表列表失败: {str(e)}")
        raise HTTPException(status_code=500, detail="获取数据表列表失败")


@router.get("/coze/tables/{table_name}/fields")
async def get_table_fields(table_name: str, db: AsyncSession = Depends(get_async_db)):
    """获取数据表的字段信息"""
    try:
        fields = await CozeService.get_table_fields(table_name, db)
        return snake_to_camel(fields)
    except Exception as e:
        logger.error(f"获取数据表字段信息失败: {str(e)}")
        raise HTTPException(status_code=500, detail="获取数据表字段信息失败")


@router.post("/coze/sync-config", response_model=dict)
async def create_sync_config(
    request: CozeUploadRequest,
    db: AsyncSession = Depends(get_async_db)
):
    """创建实时同步配置"""
    try:
        # 验证请求参数
        if not request.tableName:
            raise HTTPException(status_code=400, detail="表名不能为空")
        
        # 验证至少有一个工作流ID不为空
        if (not request.cozeWorkflowId and 
            not request.cozeWorkflowIdInsert and 
            not request.cozeWorkflowIdUpdate and 
            not request.cozeWorkflowIdDelete):
            raise HTTPException(status_code=400, detail="至少需要设置一个工作流ID（新增、更新或删除操作）")
        
        # 创建实时同步配置
        config_id = await CozeService.create_sync_config(
            table_name=request.tableName,
            coze_workflow_id=request.cozeWorkflowId,
            coze_workflow_id_insert=request.cozeWorkflowIdInsert,
            coze_workflow_id_update=request.cozeWorkflowIdUpdate,
            coze_workflow_id_delete=request.cozeWorkflowIdDelete,
            coze_api_url=request.cozeApiUrl,
            coze_api_key=request.cozeApiKey,  # 直接传递API密钥（服务层已支持大驼峰命名）
            selected_fields=request.selectedFields,  # 传递选择的字段列表
            config_title=request.configTitle,  # 传递配置标题
            db=db  # 传递数据库会话
        )
        
        # 记录操作日志
        # 注意：OperationLogService使用create_log方法，不是create_operation_log
        # 这里需要传入数据库会话和完整的日志数据对象
        # 由于当前没有数据库会话，暂时注释掉操作日志记录
        # await OperationLogService.create_log(db, log_data)
        
        return {
            "configId": config_id,
            "message": "实时同步配置创建成功",
            "status": "CREATED"
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"创建实时同步配置失败: {str(e)}")
        
        # 记录失败的操作日志
        # 注意：OperationLogService使用create_log方法，不是create_operation_log
        # 这里需要传入数据库会话和完整的日志数据对象
        # 由于当前没有数据库会话，暂时注释掉操作日志记录
        # await OperationLogService.create_log(db, log_data)
        
        raise HTTPException(status_code=500, detail="创建实时同步配置失败")

@router.get("/coze/sync-configs", response_model=CozeSyncConfigListResponse)
async def get_sync_configs(db: AsyncSession = Depends(get_async_db)):
    """获取所有同步配置"""
    try:
        configs = await CozeService.get_sync_configs(db)
        # 应用命名转换，确保返回数据符合前端期望的命名格式
        configs_camel = snake_to_camel(configs)
        
        # 转换为CozeSyncConfigListResponse格式
        return CozeSyncConfigListResponse(
            items=configs_camel,
            total=len(configs_camel)
        )
        
    except Exception as e:
        logger.error(f"获取同步配置失败: {str(e)}")
        raise HTTPException(status_code=500, detail="获取同步配置失败")

@router.put("/coze/sync-config/{config_id}")
async def update_sync_config(
    config_id: str,
    updates: dict,
    db: AsyncSession = Depends(get_async_db)
):
    """更新同步配置"""
    try:
        success = await CozeService.update_sync_config(config_id, updates, db)
        if not success:
            raise HTTPException(status_code=404, detail="同步配置不存在")
        
        return {"message": "同步配置更新成功"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"更新同步配置失败: {str(e)}")
        raise HTTPException(status_code=500, detail="更新同步配置失败")

@router.delete("/coze/sync-config/{config_id}")
async def delete_sync_config(
    config_id: str,
    db: AsyncSession = Depends(get_async_db)
):
    """删除同步配置"""
    try:
        success = await CozeService.delete_sync_config(config_id, db)
        if not success:
            raise HTTPException(status_code=404, detail="同步配置不存在")
        
        return {"message": "同步配置删除成功"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"删除同步配置失败: {str(e)}")
        raise HTTPException(status_code=500, detail="删除同步配置失败")


@router.post("/coze/test-connection")
async def test_coze_connection(
    request: dict
):
    """测试Coze API连通性"""
    try:
        # 验证请求参数
        if not request.get("cozeApiUrl"):
            raise HTTPException(status_code=400, detail="API地址不能为空")
        
        # API密钥是可选的，如果没有提供则使用空字符串
        coze_api_key = request.get("cozeApiKey", "")
        
        # 测试API连接
        result = await CozeService.test_coze_connection(
            coze_api_url=request["cozeApiUrl"],
            coze_api_key=coze_api_key
        )
        
        return result
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"测试Coze API连接失败: {str(e)}")
        raise HTTPException(status_code=500, detail="测试API连接失败")


@router.get("/coze/upload/{upload_id}", response_model=CozeUploadStatus)
async def get_upload_status(
    upload_id: str
):
    """获取上传任务状态"""
    try:
        status = CozeService.get_upload_status(upload_id)
        if not status:
            raise HTTPException(status_code=404, detail="上传任务不存在")
        
        return status
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取上传状态失败: {str(e)}")
        raise HTTPException(status_code=500, detail="获取上传状态失败")


@router.get("/coze/tables/{table_name}/data")
async def get_table_data(
    table_name: str,
    limit: int = 10,
    offset: int = 0,
    db = Depends(get_async_db)
):
    """获取指定表的真实数据"""
    try:
        data = await CozeService.get_table_data(table_name, limit, offset, db)
        # 应用命名转换，确保返回数据符合前端期望的命名格式
        return snake_to_camel(data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"获取表 {table_name} 数据失败: {str(e)}")
        raise HTTPException(status_code=500, detail="获取表数据失败")


@router.get("/coze/tables/{table_name}/sample")
async def get_table_sample_data(
    table_name: str,
    sample_size: int = 5,
    db = Depends(get_async_db)
):
    """获取表的样本数据（用于测试和预览）"""
    try:
        data = await CozeService.get_table_data(table_name, limit=sample_size, db=db)
        # 应用命名转换，确保返回数据符合前端期望的命名格式
        return snake_to_camel(data)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"获取表 {table_name} 样本数据失败: {str(e)}")
        raise HTTPException(status_code=500, detail="获取表样本数据失败")


@router.get("/coze/history", response_model=List[CozeUploadHistory])
async def get_upload_history(
    page: int = 1,
    size: int = 20
):
    """获取上传历史记录"""
    try:
        history = CozeService.get_upload_history(
            page=page,
            size=size
        )
        # 应用命名转换，确保返回数据符合前端期望的命名格式
        return snake_to_camel(history)
    except Exception as e:
        logger.error(f"获取上传历史失败: {str(e)}")
        raise HTTPException(status_code=500, detail="获取上传历史失败")


@router.delete("/coze/upload/{upload_id}")
async def cancel_upload(
    upload_id: str
):
    """取消上传任务"""
    try:
        success = CozeService.cancel_upload(upload_id)
        if not success:
            raise HTTPException(status_code=404, detail="上传任务不存在或无法取消")
        
        # 记录操作日志
        # 注意：OperationLogService使用create_log方法，不是create_operation_log
        # 这里需要传入数据库会话和完整的日志数据对象
        # 由于当前没有数据库会话，暂时注释掉操作日志记录
        # await OperationLogService.create_log(db, log_data)
        
        return {"message": "上传任务已取消"}
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"取消上传任务失败: {str(e)}")
        raise HTTPException(status_code=500, detail="取消上传任务失败")


@router.post("/coze/sync-config/{config_uuid}/manual-sync")
async def trigger_manual_sync(
    config_uuid: UUID,
    db = Depends(get_async_db)
):
    """手动触发同步配置的数据同步"""
    try:
        result = await CozeService.trigger_manual_sync(config_uuid, db)
        
        # 记录操作日志
        # 注意：OperationLogService使用create_log方法，不是create_operation_log
        # 这里需要传入数据库会话和完整的日志数据对象
        # 由于当前没有数据库会话，暂时注释掉操作日志记录
        # await OperationLogService.create_log(db, log_data)
        
        return result
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"手动同步失败: {str(e)}")
        raise HTTPException(status_code=500, detail=f"手动同步失败: {str(e)}")

File: backups/xiaochuanerp_backup_20251024_093537/xiaochuancrd/frontend/src/services/api/coze.ts
import { apiClient } from './client';
import { snakeToCamel } from './mapper';

/**
 * Coze数据上传相关API
 */

// 数据表信息接口
interface CozeTableInfo {
  tableName: string;
  displayName: string;
  description: string;
  recordCount: number;
  lastUpdated?: string;
}

// 上传筛选条件接口
interface CozeUploadFilter {
  field: string;
  operator: string;
  value: any;
}

// 上传请求接口
interface CozeUploadRequest {
  tableName: string;
  cozeWorkflowId?: string;
  cozeWorkflowIdInsert?: string;
  cozeWorkflowIdUpdate?: string;
  cozeWorkflowIdDelete?: string;
  cozeApiKey?: string;
  cozeApiUrl?: string;
  filters?: CozeUploadFilter[];
  batchSize: number;
  selectedFields?: string[];
  configTitle?: string;
}

// 上传响应接口
interface CozeUploadResponse {
  uploadId: string;
  message: string;
  status: string;
}

// 上传状态接口
interface CozeUploadStatus {
  uploadId: string;
  tableName: string;
  status: string;
  progress: number;
  totalRecords: number;
  processedRecords: number;
  failedRecords: number;
  startTime?: string;
  endTime?: string;
  errorMessage?: string;
}

// 上传历史接口
interface CozeUploadHistory {
  uploadId: string;
  tableName: string;
  cozeWorkflowId: string;
  status: string;
  totalRecords: number;
  successRecords: number;
  failedRecords: number;
  startTime: string;
  endTime?: string;
  operatorName: string;
}

/**
 * Coze API服务
 */
export const cozeApi = {
  /**
   * 获取可上传的数据表列表
   */
  async getTables(): Promise<{
    success: boolean;
    data: CozeTableInfo[];
    message?: string;
  }> {
    try {
      const response = await apiClient.request('/api/v1/coze/tables', {
        method: 'GET',
      });
      
      return {
        success: true,
        data: response,
      };
    } catch (error: any) {
      console.error('获取数据表列表失败:', error);
      return {
        success: false,
        data: [],
        message: error.message || '获取数据表列表失败',
      };
    }
  },

  /**
   * 获取数据表的字段信息
   */
  async getTableFields(tableName: string): Promise<{
    success: boolean;
    data: any[];
    message?: string;
  }> {
    try {
      const response = await apiClient.request(`/api/v1/coze/tables/${tableName}/fields`, {
        method: 'GET',
      });
      
      return {
        success: true,
        data: response,
      };
    } catch (error: any) {
      console.error('获取数据表字段信息失败:', error);
      return {
        success: false,
        data: [],
        message: error.message || '获取数据表字段信息失败',
      };
    }
  },

  /**
   * 创建实时同步配置
   */
  async createSyncConfig(request: CozeUploadRequest): Promise<{
    success: boolean;
    data?: {
      configId: string;
      message: string;
      status: string;
    };
    message?: string;
  }> {
    try {
      const response = await apiClient.request('/api/v1/coze/sync-config', {
        method: 'POST',
        body: request,
      });
      
      return {
        success: true,
        data: response,
      };
    } catch (error: any) {
      console.error('创建同步配置失败:', error);
      return {
        success: false,
        message: error.message || '创建同步配置失败',
      };
    }
  },

  /**
   * 获取所有同步配置
   */
  async getSyncConfigs(): Promise<{
    success: boolean;
    data?: any[];
    message?: string;
  }> {
    try {
      const response = await apiClient.request('/api/v1/coze/sync-configs', {
        method: 'GET',
      });
      
      // 将后端返回的蛇形命名数据转换为大驼峰命名
      const camelCaseData = snakeToCamel(response);
      
      // 后端返回的数据结构是 {items: [...], total: 2}，我们需要提取items数组
      return {
        success: true,
        data: camelCaseData.items || [],
      };
    } catch (error: any) {
      console.error('获取同步配置失败:', error);
      return {
        success: false,
        message: error.message || '获取同步配置失败',
      };
    }
  },

  /**
   * 更新同步配置
   */
  async updateSyncConfig(configId: string, updates: any): Promise<{
    success: boolean;
    message?: string;
  }> {
    try {
      await apiClient.request(`/api/v1/coze/sync-config/${configId}`, {
        method: 'PUT',
        body: updates,
      });
      
      return {
        success: true,
      };
    } catch (error: any) {
      console.error('更新同步配置失败:', error);
      return {
        success: false,
        message: error.message || '更新同步配置失败',
      };
    }
  },

  /**
   * 删除同步配置
   */
  async deleteSyncConfig(configId: string): Promise<{
    success: boolean;
    message?: string;
  }> {
    try {
      await apiClient.request(`/api/v1/coze/sync-config/${configId}`, {
        method: 'DELETE',
      });
      
      return {
        success: true,
      };
    } catch (error: any) {
      console.error('删除同步配置失败:', error);
      return {
        success: false,
        message: error.message || '删除同步配置失败',
      };
    }
  },

  /**
   * 获取上传任务状态
   */
  async getUploadStatus(uploadId: string): Promise<{
    success: boolean;
    data?: CozeUploadStatus;
    message?: string;
  }> {
    try {
      const response = await apiClient.request(`/api/v1/coze/upload/${uploadId}`, {
        method: 'GET',
      });
      
      return {
        success: true,
        data: response,
      };
    } catch (error: any) {
      console.error('获取上传状态失败:', error);
      return {
        success: false,
        message: error.message || '获取上传状态失败',
      };
    }
  },

  /**
   * 获取上传历史记录
   */
  async getUploadHistory(page: number = 1, size: number = 20): Promise<{
    success: boolean;
    data: CozeUploadHistory[];
    message?: string;
  }> {
    try {
      const response = await apiClient.request('/api/v1/coze/history', {
        method: 'GET',
        params: { page, size },
      });
      
      return {
        success: true,
        data: response,
      };
    } catch (error: any) {
      console.error('获取上传历史失败:', error);
      return {
        success: false,
        data: [],
        message: error.message || '获取上传历史失败',
      };
    }
  },

  /**
   * 取消上传任务
   */
  async cancelUpload(uploadId: string): Promise<{
    success: boolean;
    message?: string;
  }> {
    try {
      await apiClient.request(`/api/v1/coze/upload/${uploadId}`, {
        method: 'DELETE',
      });
      
      return {
        success: true,
      };
    } catch (error: any) {
      console.error('取消上传任务失败:', error);
      return {
        success: false,
        message: error.message || '取消上传任务失败',
      };
    }
  },

  /**
   * 测试Coze API连接
   */
  async testConnection(cozeApiUrl: string, cozeApiKey: string): Promise<{
    success: boolean;
    message?: string;
    statusCode?: number;
  }> {
    try {
      const response = await apiClient.request('/api/v1/coze/test-connection', {
        method: 'POST',
        body: {
          cozeApiUrl,
          cozeApiKey
        }
      });
      
      return response;
    } catch (error: any) {
      console.error('Coze API连接测试失败:', error);
      return {
        success: false,
        message: error.message || 'Coze API连接测试失败',
      };
    }
  },

  /**
   * 验证Coze工作流ID
   */
  async validateWorkflow(_workflowId: string, _apiKey?: string): Promise<{
    success: boolean;
    workflowName?: string;
    message?: string;
  }> {
    try {
      // 这里实现Coze工作流验证逻辑
      // 暂时返回成功
      return {
        success: true,
        workflowName: '示例工作流',
        message: '工作流验证成功',
      };
    } catch (error: any) {
      console.error('工作流验证失败:', error);
      return {
        success: false,
        message: error.message || '工作流验证失败',
      };
    }
  },

  /**
   * 手动触发同步配置的数据同步
   */
  async triggerManualSync(configUuid: string): Promise<{
    success: boolean;
    message?: string;
    recordsSynced?: number;
    uploadId?: string;
  }> {
    try {
      const response = await apiClient.request(`/api/v1/coze/sync-config/${configUuid}/manual-sync`, {
        method: 'POST',
      });
      
      return {
        success: response.success,
        message: response.message,
        recordsSynced: response.records_synced,
        uploadId: response.upload_id,
      };
    } catch (error: any) {
      console.error('手动同步失败:', error);
      return {
        success: false,
        message: error.message || '手动同步失败',
      };
    }
  },
};

/**
 * Coze配置相关API
 */
export const cozeConfigApi = {
  /**
   * 获取Coze API配置
   */
  async getConfig(): Promise<{
    success: boolean;
    data?: {
      apiKey?: string;
      baseUrl: string;
      timeout: number;
    };
    message?: string;
  }> {
    try {
      // 这里可以从本地存储或后端获取配置
      const config = {
        baseUrl: 'https://api.coze.cn',
        timeout: 30,
      };
      
      return {
        success: true,
        data: config,
      };
    } catch (error: any) {
      console.error('获取Coze配置失败:', error);
      return {
        success: false,
        message: error.message || '获取Coze配置失败',
      };
    }
  },

  /**
   * 保存Coze API配置
   */
  async saveConfig(config: {
    apiKey: string;
    baseUrl?: string;
    timeout?: number;
  }): Promise<{
    success: boolean;
    message?: string;
  }> {
    try {
      // 这里可以保存配置到本地存储或后端
      localStorage.setItem('coze_api_key', config.apiKey);
      
      return {
        success: true,
        message: '配置保存成功',
      };
    } catch (error: any) {
      console.error('保存Coze配置失败:', error);
      return {
        success: false,
        message: error.message || '保存Coze配置失败',
      };
    }
  },
};